src/compute/src/extensions/operator.rs

```diff
@@ -0,0 +1,368 @@
+    }
+
+    fn mz_arrange_core<P, Tr>(&self, pact: P, name: &str) -> Arranged<G, TraceAgent<Tr>>
+    where
+        R: ExchangeData,
+        P: ParallelizationContract<G::Timestamp, ((DataflowError, ()), G::Timestamp, R)>,
+        Tr: Trace
+            + TraceReader<Key = DataflowError, Val = (), Time = G::Timestamp, R = R>
+            + 'static,
+        Tr::Batch: Batch,
+        Arranged<G, TraceAgent<Tr>>: ArrangementSize,
+    {
+        // Allow access to `arrange_named` because we're within Mz's wrapper.
+        #[allow(clippy::disallowed_methods)]
+        self.arrange_core(pact, name).log_arrangement_size()
+    }
+}
+
+// A type that can log its heap size.
+pub(crate) trait ArrangementSize {
+    fn log_arrangement_size(&self) -> Self;
+}
+
+/// Helper to compute the size of a vector in memory.
+///
+/// The function only considers the immediate allocation of the vector, but is oblivious of any
+/// pointers to owned allocations.
+#[inline]
+fn vec_size<T>(data: &Vec<T>, mut callback: impl FnMut(usize, usize)) {
+    let size_of_t = std::mem::size_of::<T>();
+    callback(data.len() * size_of_t, data.capacity() * size_of_t);
+}
+
+/// Helper for [`ArrangementSize`] to install a common operator holding on to a trace.
+fn log_arrangement_size_inner<G, Tr, L>(arranged: &Arranged<G, TraceAgent<Tr>>, mut logic: L)
+where
+    G: Scope,
+    G::Timestamp: Timestamp + Lattice + Ord,
+    Tr: TraceReader + 'static,
+    Tr::Time: Timestamp + Lattice + Ord + Clone + 'static,
+    L: FnMut(&TraceAgent<Tr>) -> (usize, usize, usize) + 'static,
+{
+    let scope = arranged.stream.scope();
+    let Some(logger) = scope.log_register().get::<ComputeEvent>("materialize/compute") else {return};
+    let mut trace = arranged.trace.clone();
+    let operator = trace.operator().global_id;
+
+    // We don't want to block compaction.
+    trace.set_logical_compaction(Antichain::new().borrow());
+    trace.set_physical_compaction(Antichain::new().borrow());
+
+    let (mut old_size, mut old_capacity, mut old_count) = (0isize, 0isize, 0isize);
+
+    let mut builder = OperatorBuilder::new("ArrangementSize".to_owned(), scope);
+    let mut input = builder.new_input(&arranged.stream, Pipeline);
+    let address = builder.operator_info().address;
+    logger.log(ComputeEvent::ArrangementHeapSizeOperator { operator, address });
+    builder.build(|_cap| {
+        move |frontiers| {
+            input.for_each(|_time, _data| {});
+            if frontiers[0].is_empty() {
+                // TODO(mh): Correctness
+                // This is not correct: We can maintain a trace after the arrangement operator
+                // has been dropped, which means that we report that the trace does not consume
+                // space, but it might still do so. Fixing it would require Differential logging
+                // to signal when trace sharing drops to 0 (or 1 because this operator shares one
+                // copy.)
```

Comment:
@frankmcsherry  had some ideas for how to deal with this issue, but maybe it doesn't matter.

Reply:
I changed it such that it'll retract the information once the trace isn't shared anymore.